### PR TITLE
fix: standardise logging format to use json 

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* standardise logging format to use JSON
* moved logging setup into it's own package to make it easier to use across other packages
* new logging format for auth events will follow:
```bash
{"auth_event":{"action":"user_signedup","actor_id":"c93b6eb0-4af8-4247-84a9-00c88c041b75","actor_username":"foo@example.com","log_type":"team"},"component":"api","duration":386683041,"level":"info","method":"GET","msg":"request completed","path":"/callback","referer":"","remote_addr":"127.0.0.1:54356","status":302,"timestamp":"2022-07-23T08:33:22Z"}

```

### CODE REVIEW

File: api/external.go [hunk 1]
```diff
@@ -260,6 +261,14 @@ func (a *API) internalExternalProviderCallback(w http.ResponseWriter, r *http.Re
 				}); terr != nil {
 					return terr
 				}
+				logger.LogEntrySetFields(r, logrus.Fields{
```
Comment (kangmingtay) [inline api/external.go:264]: not too sure how we can reduce duplicated code here... ideally we'll want to include the auth event in the logs everytime there is a new row insert into the `audit_log_entries` table

another solution i thought of was to pass the `http.Request` as an argument into `NewAuditLogEntry` (https://github.com/supabase/gotrue/pull/562/files#diff-57bd73e87d7b244be8c43b1e9acf055f2bf61b9f714cffa760298b3314bce6e6R66) so we can just set the payload in the log fields using `logger.LogEntrySetFields(request, payload) - would this be a cleaner approach to reduce code duplication?

Comment (kangmingtay) [reply api/external.go:264]: cc @hf @J0

**Comment by kangmingtay** [comment]
@chasers lmk if this is easier for you to parse - we can also change the logging key from `auth_event` to something else to make it more explicit if you want

**Review by chasers — APPROVED** [review]
As far as the format goes ... looks great. Will be much cleaner to handle down the line.

**Comment by chasers** [comment]
@burmecia FYI

File: logger/log.go [hunk 2]
```diff
@@ -25,10 +28,11 @@ func (l *structuredLogger) NewLogEntry(r *http.Request) chimiddleware.LogEntry {
 		"path":        r.URL.Path,
 		"remote_addr": r.RemoteAddr,
 		"referer":     r.Referer(),
+		"timestamp":   time.Now().UTC().Format(time.RFC3339),
 	}
 
-	if reqID := getRequestID(r.Context()); reqID != "" {
-		logFields["request_id"] = reqID
+	if reqID := r.Context().Value("request_id"); reqID != nil {
```
Comment (J0) [inline logger/log.go:34]: Clarification: what's stopping us  from using `getRequestID()` ? 

Would it be to avoid having a circular import?

**Review by J0 — APPROVED** [review]
Looks good to me, thanks for the PR!

File: api/external.go (hunk 1 above)
Comment (J0) [reply api/external.go:264]: Not sure if it's a permissions thing but can't quite see the Github diff: https://github.com/supabase/gotrue/pull/562/files#diff-57bd73e87d7b244be8c43b1e9acf055f2bf61b9f714cffa760298b3314bce6e6R66 

Am OK with either solution  --   passing in `http.Request` to `NewAuditLogEntry` or having code duplication. If we  really wanted to try and cut back on code I suppose we could use a function to generate the auth event struct

```go
func generateEvent(actorID, actorUsername .....) {
                         return logrus.Fields{
 						"actor_id":       actorID,
 						"actor_username": user.GetEmail(),
 						"action":         models.UserSignedUpAction,
 						"log_type":       models.ActionLogTypeMap[models.UserSignedUpAction],
 					}

}
```

will defer to @hf  here

File: logger/log.go (hunk 2 above)
Comment (kangmingtay) [reply logger/log.go:34]: it's a private function so it's only accessible in the `api` package

Comment (J0) [reply logger/log.go:34]: ah ok missed the lower case. Thanks for the clarification!

**Review by hf — APPROVED** [review]

File: api/external.go (hunk 1 above)
Comment (kangmingtay) [reply api/external.go]: went with passing the request into `NewAuditLogEntry` so we don't have to add `logger.LogEntrySetFields` everywhere

**Comment by kangmingtay** [comment]
i've also removed the timestamp field from `auth.audit_log_entries->>'payload'` because there's already a column called `created_at` which is basically the same thing

**Pull request merged**